h5py/_hl/base.py

```diff
@@ -120,7 +120,7 @@ def array_for_new_object(data, specified_dtype=None):
     # In most cases, this does nothing. But if data was already an array,
     # and as_dtype is a tagged h5py dtype (e.g. for an object array of strings),
     # asarray() doesn't replace its dtype object. This gives it the tagged dtype:
-    if as_dtype is not None:
+    if as_dtype is not None and data.dtype.kind != "T":
```

Comment:
Maybe I missed it, but why are we special-casing the variable width strings here?

Reply:
There's some weird behaviour with literal string `'T'` dtype. I reworked it by ensuring it's a np.dtype object.